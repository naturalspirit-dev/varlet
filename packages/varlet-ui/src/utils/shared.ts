export interface CacheInstance<T> {
	cache: T[]

	has(key: T): boolean

	add(key: T): void

	remove(key: T): void
}

export const isString = (val: unknown): val is string => typeof val === 'string'

export const isNumber = (val: unknown): val is number => typeof val === 'number'

export const isBaseObject = (val: unknown): val is Record<string, any> =>
  Object.prototype.toString.call(val) === '[object Object]'

export const isArray = (val: unknown): val is Array<any> => Array.isArray(val)

export const isURL = (val: string) => /^(http)|(\.*\/)/.test(val)

export const isEmpty = (val: unknown) =>
  val === undefined || val === null || val === '' || (Array.isArray(val) && !val.length)

export const removeItem = (arr: Array<unknown>, item: unknown) => {
  if (arr.length) {
    const index: number = arr.indexOf(item)
    if (index > -1) {
      return arr.splice(index, 1)
    }
  }
}

export const throttle = (method: any, mustRunDelay = 200): (() => void) => {
  let timer: number
  let start: number
  return function loop(this: unknown, ...args) {
    const now = Date.now()
    if (!start) {
      start = now
    }
    if (timer) {
      window.clearTimeout(timer)
    }
    if (now - start >= mustRunDelay) {
      method.apply(this, args)
      start = now
    } else {
      timer = window.setTimeout(() => {
        loop.apply(this, args)
      }, 50)
    }
  }
}

export const debounce = (method: any, delay = 200) => {
  let timer: number
  return function (this: unknown, ...args: any[]) {
    timer && window.clearTimeout(timer)
    timer = window.setTimeout(() => {
      method.apply(this, args)
    }, delay)
  }
}

export const checkIntersectionObserverAPI = (): boolean => 'IntersectionObserver' in window

export const createInViewportObserver = <T>(handler: (el: T) => void): IntersectionObserver => {
  return new IntersectionObserver((intersectionObserverEntries: IntersectionObserverEntry[]) => {
    intersectionObserverEntries.forEach((intersectionObserverEntry: IntersectionObserverEntry) => {
      if (intersectionObserverEntry.intersectionRatio > 0) {
        handler((intersectionObserverEntry.target as unknown) as T)
      }
    })
  })
}

export const createCache = <T>(max: number): CacheInstance<T> => {
  const cache: T[] = []

  return {
    cache,
    has(key: T) {
      return cache.includes(key)
    },
    add(key: T) {
      if (this.has(key)) {
        return
      }

      this.cache.length === max && cache.shift()
      this.cache.push(key)
    },
    remove(key: T) {
      this.has(key) && removeItem(this.cache, key)
    },
  }
}
