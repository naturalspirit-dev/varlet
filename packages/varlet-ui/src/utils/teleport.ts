import { onActivated, onDeactivated, ref, Ref } from 'vue'

export function useTeleport() {
  const disabled: Ref<boolean> = ref(false)
  onActivated(() => {
    disabled.value = false
  })
  onDeactivated(() => {
    disabled.value = true
  })

  return {
    disabled
  }
}
