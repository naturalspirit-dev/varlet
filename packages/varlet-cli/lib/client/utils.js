import { kebabCase } from '@varlet/shared';
import { onMounted, onUnmounted } from 'vue';
import { get } from 'lodash-es';
import config from '@config';
const mountedVarKeys = [];
export function StyleProvider(styleVars = {}) {
    mountedVarKeys.forEach((key) => document.documentElement.style.removeProperty(key));
    mountedVarKeys.length = 0;
    const styles = formatStyleVars(styleVars);
    Object.entries(styles).forEach(([key, value]) => {
        document.documentElement.style.setProperty(key, value);
        mountedVarKeys.push(key);
    });
}
function formatStyleVars(styleVars) {
    return Object.entries(styleVars !== null && styleVars !== void 0 ? styleVars : {}).reduce((styles, [key, value]) => {
        const cssVar = key.startsWith('--') ? key : `--${kebabCase(key)}`;
        styles[cssVar] = value;
        return styles;
    }, {});
}
export function getPCLocationInfo() {
    const [, language, menuName] = window.location.hash.split('/');
    return {
        language,
        menuName,
    };
}
export function addRouteListener(cb) {
    onMounted(() => {
        window.addEventListener('hashchange', cb);
        window.addEventListener('popstate', cb);
    });
    onUnmounted(() => {
        window.removeEventListener('hashchange', cb);
        window.removeEventListener('popstate', cb);
    });
}
export function getHashSearch() {
    const { href } = window.location;
    const hashSearch = href.slice(href.indexOf('?'));
    return new URLSearchParams(hashSearch);
}
export function getBrowserTheme() {
    var _a;
    const themeKey = get(config, 'themeKey');
    const darkThemeConfig = get(config, 'darkTheme');
    if (!darkThemeConfig) {
        return 'lightTheme';
    }
    const storageTheme = window.localStorage.getItem(themeKey);
    if (!storageTheme) {
        const preferTheme = ((_a = window.matchMedia) === null || _a === void 0 ? void 0 : _a.call(window, '(prefers-color-scheme: dark)').matches) ? 'darkTheme' : 'lightTheme';
        window.localStorage.setItem(themeKey, preferTheme);
        return preferTheme;
    }
    return storageTheme;
}
