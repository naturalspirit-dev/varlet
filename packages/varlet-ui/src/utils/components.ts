import {
	createApp,
	Component,
	h,
	getCurrentInstance,
	inject,
	onUnmounted,
	computed,
	ComputedRef,
	provide,
	reactive,
	VNode,
	isVNode,
	ComponentInternalInstance,
	onMounted,
	onBeforeUnmount,
	nextTick,
} from 'vue'
import { removeItem } from './shared'

export interface MountInstance {
	instance: any
	unmount: () => void
}

export interface ChildrenCounter {
	collect(instance: ComponentInternalInstance): void
	clear(instance: ComponentInternalInstance): void
	instances: ComponentInternalInstance[]
}

export interface BaseParentProvider<C> {
	collect(childProvider: C): void
	clear(childProvider: C): void
}

export function pickProps(props: any, propsKey: string): any
export function pickProps(props: any, propsKey: string[]): any
export function pickProps(props: any, propsKey: any): any {
	return Array.isArray(propsKey)
		? propsKey.reduce((pickedProps: any, key) => {
				pickedProps[key] = props[key]
				return pickedProps
		  }, {})
		: props[propsKey]
}

export function mount(component: Component): MountInstance {
	const app = createApp(component)
	const host = document.createElement('div')
	document.body.appendChild(host)

	return {
		instance: app.mount(host),
		unmount() {
			app.unmount(host)
			document.body.removeChild(host)
		},
	}
}

export function mountInstance(
	component: any,
	props: Record<string, any> = {},
	eventListener: Record<string, any> = {}
): {
	unmountInstance: () => void
} {
	const Host = {
		setup() {
			return () =>
				h(component, {
					...props,
					...eventListener,
				})
		},
	}

	const { unmount } = mount(Host)
	return { unmountInstance: unmount }
}

export function flatVNodes(subTree: any) {
	const vNodes: VNode[] = []

	const flat = (subTree: any) => {
		if (subTree?.component) {
			flat(subTree?.component.subTree)
			return
		}

		if (Array.isArray(subTree?.children)) {
			subTree.children.forEach((child: any) => {
				if (isVNode(child)) {
					vNodes.push(child)

					flat(child)
				}
			})
		}
	}

	flat(subTree)

	return vNodes
}

export function useAtChildrenCounter(key: symbol) {
	const instances: ComponentInternalInstance[] = reactive([])
	const parentInstance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

	const sortInstances = () => {
		const vNodes: any[] = flatVNodes(parentInstance.subTree)

		instances.sort((a, b) => {
			return vNodes.indexOf(a.vnode) - vNodes.indexOf(b.vnode)
		})
	}

	const collect = (instance: ComponentInternalInstance) => {
		instances.push(instance)
		sortInstances()
	}

	const clear = (instance: ComponentInternalInstance) => {
		removeItem(instances, instance)
	}

	provide<ChildrenCounter>(key, {
		collect,
		clear,
		instances,
	})

	const length: ComputedRef<number> = computed(() => instances.length)

	return {
		length,
	}
}

export function useAtParentIndex(key: symbol) {
	const childrenCounter: ChildrenCounter | undefined = inject<ChildrenCounter>(key)
	if (!childrenCounter) {
		throw new Error('Children counter provider not found')
	}

	const { collect, clear, instances } = childrenCounter

	const instance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

	nextTick().then(() => collect(instance))
	onUnmounted(() => clear(instance))

	const index = computed(() => instances.indexOf(instance))

	return {
		index,
	}
}

export function useChildren<P, C>(key: symbol) {
	const childProviders: C[] = []

	const collect = (childProvider: C) => {
		childProviders.push(childProvider)
	}

	const clear = (childProvider: C) => {
		removeItem(childProviders, childProvider)
	}

	const bindChildren = (parentProvider: P) => {
		provide<P & BaseParentProvider<C>>(key, {
			collect,
			clear,
			...parentProvider,
		})
	}

	return {
		childProviders,
		bindChildren,
	}
}

export function useParent<P, C>(key: symbol) {
	const { collect, clear, ...parentProvider } = inject<P & BaseParentProvider<C>>(key) as P & BaseParentProvider<C>
	if (!collect || !clear) {
		throw new Error('Parent not found')
	}

	const bindParent = (childProvider: C) => {
		onMounted(() => collect(childProvider))
		onBeforeUnmount(() => clear(childProvider))
	}

	return {
		parentProvider,
		bindParent,
	}
}
