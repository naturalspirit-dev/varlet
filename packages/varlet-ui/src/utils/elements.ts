import { isNumber, isString, toNumber } from './shared'

export function getTop(element: HTMLElement): number {
  const { top } = element.getBoundingClientRect()

  return top + (document.body.scrollTop || document.documentElement.scrollTop)
}

export function getScrollTop(element: Element | Window): number {
  const top = 'scrollTop' in element ? element.scrollTop : element.pageYOffset

  // iOS scroll bounce cause minus scrollTop
  return Math.max(top, 0)
}

export function getScrollLeft(element: Element | Window): number {
  const left = 'scrollLeft' in element ? element.scrollLeft : element.pageXOffset

  return Math.max(left, 0)
}

export function getLeft(element: HTMLElement): number {
  const { left } = element.getBoundingClientRect()
  return left + (document.body.scrollLeft || document.documentElement.scrollLeft)
}

export function inViewport(element: HTMLElement): boolean {
  const { top, bottom, left, right } = element.getBoundingClientRect()

  const xInViewport = left < window.innerWidth && right > 0
  const yInViewport = top < window.innerHeight && bottom > 0

  return xInViewport && yInViewport
}

export function getTranslate(el: HTMLElement) {
  const { transform } = window.getComputedStyle(el)
  return +transform.slice(transform.lastIndexOf(',') + 2, transform.length - 1)
}

export function isHidden(el: HTMLElement) {
  const { width, height } = el.getBoundingClientRect()

  return width === 0 && height === 0
}

export function getParentScroller(el: HTMLElement): HTMLElement | Window {
  let element = el

  while (element) {
    if (!element.parentNode) {
      break
    }

    element = element.parentNode as HTMLElement

    if (element === document.body || element === document.documentElement) {
      break
    }

    const scrollRE = /(scroll|auto)/
    const { overflowY } = window.getComputedStyle(element)

    if (scrollRE.test(overflowY)) {
      return element
    }
  }

  return window
}

// example 1rem
export const isRem = (value: unknown) => isString(value) && value.endsWith('rem')

// example 1 || 1px
export const isPx = (value: unknown) => (isString(value) && value.endsWith('px')) || isNumber(value)

// example 1%
export const isPercent = (value: unknown) => isString(value) && value.endsWith('%')

// example 1vw
export const isVw = (value: unknown) => isString(value) && value.endsWith('vw')

// example 1vh
export const isVh = (value: unknown) => isString(value) && value.endsWith('vh')

// example return 1
export const toPxNum = (value: unknown): number => {
  if (isNumber(value)) {
    return value
  }

  if (isPx(value)) {
    return +(value as string).replace('px', '')
  }

  if (isVw(value)) {
    return (+(value as string).replace('vw', '') * window.innerWidth) / 100
  }

  if (isVh(value)) {
    return (+(value as string).replace('vh', '') * window.innerHeight) / 100
  }

  if (isRem(value)) {
    const num = +(value as string).replace('rem', '')
    const rootFontSize = window.getComputedStyle(document.documentElement).fontSize

    return num * parseFloat(rootFontSize)
  }

  if (isString(value)) {
    return toNumber(value)
  }

  // % and other
  return 0
}

// example return 1px 1% 1vw 1vh 1rem null
export const toSizeUnit = (value: unknown) => {
  if (value == null) {
    return null
  }

  if (isPercent(value) || isVw(value) || isVh(value) || isRem(value)) {
    return value
  }

  return `${toPxNum(value)}px`
}

export function requestAnimationFrame(fn: FrameRequestCallback): number {
  return window.requestAnimationFrame ? window.requestAnimationFrame(fn) : window.setTimeout(fn, 16)
}

export function cancelAnimationFrame(handle: number): void {
  window.cancelAnimationFrame ? window.cancelAnimationFrame(handle) : window.clearTimeout(handle)
}

export function nextTickFrame(fn: FrameRequestCallback) {
  requestAnimationFrame(() => {
    requestAnimationFrame(fn)
  })
}

interface ScrollToOptions {
  top?: number
  left?: number
  duration?: number
  animation: (progress: number) => number
}

export function scrollTo(
  element: HTMLElement | Window,
  { top = 0, left = 0, duration = 300, animation }: ScrollToOptions
) {
  const startTime = Date.now()

  const scrollTop = getScrollTop(element)
  const scrollLeft = getScrollLeft(element)

  const frame = () => {
    const progress = (Date.now() - startTime) / duration

    if (progress < 1) {
      const nextTop = scrollTop + (top - scrollTop) * animation(progress)
      const nextLeft = scrollLeft + (left - scrollLeft) * animation(progress)

      element.scrollTo(nextLeft, nextTop)
      requestAnimationFrame(frame)
    } else {
      element.scrollTo(left, top)
    }
  }

  requestAnimationFrame(frame)
}
