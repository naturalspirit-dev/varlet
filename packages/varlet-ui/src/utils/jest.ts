import { nextTick } from 'vue'
import type { ComponentPublicInstance } from 'vue'
import { config } from '@vue/test-utils'
import type { VueWrapper, DOMWrapper } from '@vue/test-utils'
import { isPlainObject } from './shared'

export const delay = (time: number) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })

  el.dispatchEvent(event)

  return nextTick()
}

export function mockOffset() {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return parseFloat(window.getComputedStyle(this).width) || 0
      },
    },
    offsetHeight: {
      get() {
        return parseFloat(window.getComputedStyle(this).height) || 0
      },
    },
    clientWidth: {
      get() {
        return parseFloat(window.getComputedStyle(this).width) || 0
      },
    },
    clientHeight: {
      get() {
        return parseFloat(window.getComputedStyle(this).height) || 0
      },
    },
    offsetLeft: {
      get() {
        return parseFloat(window.getComputedStyle(this).marginLeft) || 0
      },
    },
    offsetTop: {
      get() {
        return parseFloat(window.getComputedStyle(this).marginTop) || 0
      },
    },
    offsetParent: {
      get() {
        return this.parentNode ?? {}
      },
    },
  })
}

interface Entry {
  intersectionRatio: number
  target: null | HTMLElement
}

const entry: Entry = {
  intersectionRatio: 0,
  target: null,
}

let callback: any

export function mockIntersectionObserver() {
  const originMethod = window.IntersectionObserver

  const IntersectionObserver = function (cb: any) {
    callback = cb
  }
  IntersectionObserver.prototype.observe = (el: HTMLElement) => {
    entry.target = el
  }
  IntersectionObserver.prototype.unobserve = (el: HTMLElement) => {
    entry.target = null
  }

  Object.assign(window, {
    IntersectionObserver,
  })

  return {
    mockRestore() {
      entry.intersectionRatio = 0
      entry.target = null
      window.IntersectionObserver = originMethod
    },
    trigger(intersectionRatio: number) {
      entry.intersectionRatio = intersectionRatio
      callback?.([entry])
      return entry.target
    },
  }
}

export async function triggerDrag(
  el: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | HTMLElement,
  x = 0,
  y = 0
) {
  await trigger(el, 'touchstart', 0, 0)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await trigger(el, 'touchend', x, y)
}

export function mockTranslate() {
  const originMethod = window.getComputedStyle

  const XRE = /translateX\((\d+)px\)/
  const YRE = /translateY\((\d+)px\)/

  Object.assign(window, {
    getComputedStyle: (el: HTMLElement) => {
      const styles = originMethod.call(window, el)
      const x = styles.transform.match(XRE)?.[1] ?? 0
      const y = styles.transform.match(YRE)?.[1] ?? 0
      styles.transform = `matrix(1, 0, 0, 1, ${x}, ${y})`
      return styles
    },
  })

  return {
    mockRestore() {
      window.getComputedStyle = originMethod
    },
  }
}

interface MockFileReaderCtx {
  result: string
  onload: () => void
  readAsDataURL: () => void
}

export function mockFileReader(url: string) {
  const originMethod = window.FileReader

  Object.assign(window, {
    FileReader(this: MockFileReaderCtx) {
      this.onload = function () {}
      this.result = ''

      this.readAsDataURL = function () {
        this.result = url
        this.onload()
      }
    },
  })

  return {
    mockRestore() {
      window.FileReader = originMethod
    },
  }
}

export function mockStubs() {
  const originStubs = config.global.stubs

  config.global.stubs = {}

  return {
    mockRestore() {
      config.global.stubs = originStubs
    },
  }
}

export function mockConsole(method: keyof Console) {
  const originMethod = console[method]

  console[method] = () => {}

  return {
    mockRestore() {
      console[method] = originMethod
    },
  }
}

export function mockScrollTo(Element: any) {
  Element.prototype.scrollTo = function (x: ScrollToOptions | number, y?: number) {
    if (isPlainObject(x)) {
      this.scrollLeft = x.left
      this.scrollTop = x.top
    } else {
      this.scrollLeft = x
      this.scrollTop = y
    }
  }
}
