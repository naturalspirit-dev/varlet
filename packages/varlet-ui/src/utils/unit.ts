import { isNumber, isString } from './shared'


