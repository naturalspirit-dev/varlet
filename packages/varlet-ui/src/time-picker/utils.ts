import { hours24, hoursAmpm } from './props'
import type { AmPm } from './props'
import { toNumber } from '../utils/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    return maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    return minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    return (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }
  return false
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    return (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    return (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    return (
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
    )
  }
  return false
}
