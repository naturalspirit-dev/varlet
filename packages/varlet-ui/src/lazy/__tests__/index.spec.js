import example from '../example'
import Lazy, { imageCache } from '..'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { mockIntersectionObserver, trigger } from '../../utils/jest'

test('test lazy example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test lazy plugin', () => {
  const app = createApp({}).use(Lazy)
  expect(app.directive('lazy')).toBeTruthy()
})

const Wrapper = {
  directives: { Lazy },
  data: () => ({
    src: 'https://varlet.gitee.io/varlet-ui/cat.jpg',
    error: 'https://varlet.gitee.io/varlet-ui/error.jpg',
  }),
  template: `
    <img
      :lazy-error="error"
      v-lazy="src"
    >
  `,
}

test('test lazy load use io', async () => {
  const { mockRestore, trigger: triggerCheck } = mockIntersectionObserver()
  const wrapper = mount(Wrapper)
  expect(wrapper.html()).toMatchSnapshot()

  const el = triggerCheck(1)
  await trigger(el._lazy.preloadImage, 'load')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
  imageCache.clear()
  mockRestore()
})

test('test lazy error use io', async () => {
  const { mockRestore, trigger: triggerCheck } = mockIntersectionObserver()
  const wrapper = mount(Wrapper)
  expect(wrapper.html()).toMatchSnapshot()

  const el = triggerCheck(1)
  await trigger(el._lazy.preloadImage, 'error')
  await trigger(el._lazy.preloadImage, 'error')
  await trigger(el._lazy.preloadImage, 'error')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
  imageCache.clear()
  mockRestore()
})

test('test lazy updated use io', async () => {
  const { mockRestore, trigger: triggerCheck } = mockIntersectionObserver()
  const wrapper = mount(Wrapper)
  expect(wrapper.html()).toMatchSnapshot()

  let el = triggerCheck(1)
  await trigger(el._lazy.preloadImage, 'load')
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setData({ src: 'https://varlet.gitee.io/varlet-ui/dog.jpg' })
  el = triggerCheck(1)
  await trigger(el._lazy.preloadImage, 'load')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
  imageCache.clear()
  mockRestore()
})

test('test lazy background-image load use io', async () => {
  const { mockRestore, trigger: triggerCheck } = mockIntersectionObserver()
  const wrapper = mount({
    directives: { Lazy },
    template: `
      <img v-lazy:background-image="'https://varlet.gitee.io/varlet-ui/cat.jpg'">
    `,
  })
  expect(wrapper.html()).toMatchSnapshot()

  const el = triggerCheck(1)
  await trigger(el._lazy.preloadImage, 'load')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
  imageCache.clear()
  mockRestore()
})
