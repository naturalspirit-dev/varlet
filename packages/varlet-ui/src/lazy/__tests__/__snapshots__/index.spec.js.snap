// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test lazy background-image load use io 1`] = `"<img style=\\"background-image: url(data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==);\\">"`;

exports[`test lazy background-image load use io 2`] = `"<img style=\\"background-image: url(https://varlet.gitee.io/varlet-ui/cat.jpg);\\">"`;

exports[`test lazy error use io 1`] = `"<img lazy-error=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\">"`;

exports[`test lazy error use io 2`] = `"<img lazy-error=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\" src=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\">"`;

exports[`test lazy example 1`] = `
"<div class=\\"app-type\\">基本使用</div><img class=\\"cat\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\"><img class=\\"cat\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\"><img class=\\"cat\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\"><img class=\\"cat\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\"><img class=\\"cat\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\">
<div class=\\"app-type\\">背景图懒加载</div>
<div class=\\"cat\\" style=\\"background-image: url(data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==);\\"></div>"
`;

exports[`test lazy load use io 1`] = `"<img lazy-error=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\">"`;

exports[`test lazy load use io 2`] = `"<img lazy-error=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\" src=\\"https://varlet.gitee.io/varlet-ui/cat.jpg\\">"`;

exports[`test lazy updated use io 1`] = `"<img lazy-error=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\\">"`;

exports[`test lazy updated use io 2`] = `"<img lazy-error=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\" src=\\"https://varlet.gitee.io/varlet-ui/cat.jpg\\">"`;

exports[`test lazy updated use io 3`] = `"<img lazy-error=\\"https://varlet.gitee.io/varlet-ui/error.jpg\\" src=\\"https://varlet.gitee.io/varlet-ui/dog.jpg\\">"`;
